### PR TITLE
rec: Sanitize records received before doing anything else

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

Stricter version of https://github.com/PowerDNS/pdns/pull/7258

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [x] documented the code
- [x] added or modified regression test(s)
- [x] added or modified unit test(s)

### CODE REVIEW

File: pdns/syncres.cc [hunk 1]
```diff
@@ -1969,6 +1969,162 @@ vState SyncRes::validateRecordsWithSigs(unsigned int depth, const DNSName& qname
...
+    }
+
+    /* dealing with the records in additional */
+    if (rec->d_place == DNSResourceRecord::ADDITIONAL && rec->d_type != QType::A && rec->d_type != QType::AAAA && rec->d_type != QType::RRSIG) {
+      LOG(prefix<<"Removing irrelevant record '"<<rec->d_name<<"|"<<DNSRecordContent::NumberToType(rec->d_type)<<"|"<<rec->d_content->getZoneRepresentation()<<"' in the "<<(int)rec->d_place<<" section received from "<<auth<<endl);
+      rec = lwr.d_records.erase(rec);
+      continue;
+    }
+
+    if (rec->d_place == DNSResourceRecord::ADDITIONAL && allowedAdditionals.count(rec->d_name) == 0) {
```
Comment (pieterlexis) [inline pdns/syncres.cc]: I _think_ we might need to check if the allowedAdditional isPartOf `auth`?

Comment (rgacogne) [reply pdns/syncres.cc]: You mean if `rec->d_name` ispartOf `auth`? If so we already checked on line 2030, unless I'm missing your point entirely.

Comment (rgacogne) [reply pdns/syncres.cc]: But how would that `A` record not be discarded by the check on line 2030? I mean sure, we would accept the `SRV` record and so add `server.jabber.example` to the `allowedAdditionals` whitelist, but the additional `A` record should be rejected for not being part of `auth` well before being checked against that whitelist, right?

Comment (pieterlexis) [reply pdns/syncres.cc]: ah, yes. You're right!

**Pull request merged**